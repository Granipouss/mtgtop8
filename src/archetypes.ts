import { getPage } from './utils';

export type ShortDeck = {
  id: number;
  eventId: number;
  name: string;
};

export type Archetype = { id: number; name: string; decks: ShortDeck[] };

const parseName = ($el: Cheerio): string =>
  $el
    .eq(1)
    .text()
    .trim()
    .replace(/ decks$/, '');

const parseDeck = ($el: Cheerio): ShortDeck => {
  const regexp = /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/;
  const match = $el.find('td').eq(1).html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse deck: ${$el.text()}`);

  const eventId = parseInt(match[1], 10);
  const id = parseInt(match[2], 10);
  const name = match[3].trim();

  return { eventId, id, name };
};

export const getArchetype = async (
  id: number,
  page = 1
): Promise<Archetype> => {
  const $ = await getPage(
    `https://www.mtgtop8.com/archetype?a=${id}&current_page=${page}`
  );

  const archetype: Archetype = {
    id,
    name: parseName($('.w_title')),
    decks: [],
  };

  $('table table[align=center] .hover_tr').each((_, el) => {
    archetype.decks.push(parseDeck($(el)));
  });

  return archetype;
};
