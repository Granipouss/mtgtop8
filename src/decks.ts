import { getPage } from './utils';

export type ShortArchetype = { name: string; id: number };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { count: number; name: string; isCompanion?: boolean };

export type Deck = {
  id: number;
  eventId: number;
  format: string;
  name: string;
  archetype: ShortArchetype;
  player: string;
  result: string;
} & {
  [group in CardGroup]: Card[];
};

const parseFormat = ($el: Cheerio): string => {
  const match = $el.html()?.match(/(.+) <div/);
  if (!match) throw new Error(`Cannot parse format: ${$el.text()}`);
  return match[1].trim();
};

const parseArchetype = ($el: Cheerio): ShortArchetype => {
  const regexp = /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/;
  const match = $el.html()?.match(regexp);
  if (!match) throw new Error(`Cannot parse archetype: ${$el.text()}`);

  const id = parseInt(match[1], 10);
  const name = match[2].trim();

  return { id, name };
};

const parseCard = ($el: Cheerio): Card => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: Card = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};

export const getDeck = async (eventId: number, id: number): Promise<Deck> => {
  const $ = await getPage(`https://www.mtgtop8.com/event?e=${eventId}&d=${id}`);

  if ($.html().includes('No event could be found.')) {
    throw new Error(`No deck ${id} for event ${eventId}`);
  }

  const deck: Deck = {
    id,
    eventId,
    name: $('table .chosen_tr :not([align=right]) .topic').text().trim(),
    player: $('table .chosen_tr [align=right] .topic').text().trim(),
    result: $('table .chosen_tr [align=center]').text().trim(),
    format: parseFormat($('table table td[align="center"]').first()),
    archetype: parseArchetype($('.Nav_link').eq(1)),
    mainboard: [],
    sideboard: [],
    commander: [],
  };

  let currentGroup: CardGroup = 'mainboard';
  $('table table table td td').each((_, el) => {
    const isCard = $('.hover_tr', el).length > 0;
    if (isCard) {
      deck[currentGroup].push(parseCard($(el)));
    } else {
      currentGroup = parseGroup($(el));
    }
  });

  return deck;
};
