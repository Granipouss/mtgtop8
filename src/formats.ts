import { getPage } from './utils';

export type Format = { name: string; code: string };

export const getFormats = async (): Promise<Format[]> => {
  const $ = await getPage(`https://www.mtgtop8.com/`);

  const formats: Format[] = [];

  $('table')
    .eq(1)
    .find('td')
    .each((_, el) => {
      const link = $(el).find('a').attr('href');
      const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
      const linkMatch = link?.match(linkRegexp);
      if (!linkMatch) return;
      const code = linkMatch[1].trim().toUpperCase();

      const alt = $(el).find('img').attr('id');
      if (!alt) return;
      const name = alt
        .replace(/^menu_/, '')
        .replace(/_/g, ' ')
        .toLowerCase();

      formats.push({ name, code });
    });

  $('#other_tooltip_content div').each((_, el) => {
    const link = $(el).find('a').attr('href');
    const linkRegexp = /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/;
    const linkMatch = link?.match(linkRegexp);
    if (!linkMatch) return;
    const code = linkMatch[1].trim().toUpperCase();

    const alt = $(el).find('img').attr('id');
    if (!alt) return;
    const name = alt
      .replace(/^menu_/, '')
      .replace(/_/g, ' ')
      .toLowerCase();

    formats.push({ name, code });
  });

  return formats;
};
