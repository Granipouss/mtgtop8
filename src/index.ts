export * from './formats';
export * from './decks';
export * from './archetypes';
