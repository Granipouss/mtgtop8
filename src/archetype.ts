import { getPage, url, match } from './utils';

import { ShortDeck } from './deck';

export type ShortArchetype = { id: number; name: string };

export class Archetype implements ShortArchetype {
  public static async get(id: number, page = 1): Promise<Archetype> {
    const $ = await getPage(url.archetype(id, page));
    return new Archetype($);
  }

  id: number;
  name: string;

  decks: ShortDeck[];

  private constructor($: CheerioStatic) {
    this.id = parseInt(match($('table table option').attr('value'), /archetype\?a=(\d+)&f=.+&meta=.+/)[0], 10);

    this.name = $('table table').eq(1).find('tr:first-child').text().replace('decks', '').trim();

    this.decks = $('table table[align=center] .hover_tr')
      .get()
      .map((el) => {
        const [eventString, deckString, name] = match(
          $(el).find('td').eq(1).html() || '',
          /^<a href="event\?e=(\d+)&amp;d=(\d+)&amp;f=MO">(.+)<\/a>$/
        );

        return {
          eventId: parseInt(eventString, 10),
          id: parseInt(deckString, 10),
          name: name.trim(),
        };
      });
  }
}
