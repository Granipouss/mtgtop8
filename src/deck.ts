import { getPage, url, match, ListOf, WithCount, parseDeckLink } from './utils';

import { ShortArchetype } from './archetype';
import { ShortFormat } from './format';

export type ShortDeck = { eventId: number; id: number; name: string };

export type CardGroup = 'mainboard' | 'sideboard' | 'commander';

export type Card = { name: string; isCompanion?: boolean };

export class Deck implements ShortDeck {
  public static async get(eventId: number, id: number): Promise<Deck> {
    const $ = await getPage(url.deck(eventId, id));

    // FIXME: check for deck also
    if ($.html().includes('No event could be found.')) {
      throw new Error(`No deck ${id} for event ${eventId}`);
    }

    return new Deck($);
  }

  eventId: number;
  id: number;
  name: string;

  format: ShortFormat;
  archetype: ShortArchetype;
  player: string;
  result: string;

  mainboard: ListOf<Card> = [];
  sideboard: ListOf<Card> = [];
  commander: ListOf<Card> = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.eventId = chosenLinkData.event;

    this.id = chosenLinkData.deck;

    this.name = $('table .chosen_tr :not([align=right]) .topic').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const [archetypeId, archetypeName] = match(
      $('.Nav_link').eq(1).html() || '',
      /^<a href="archetype\?a=(\d+)">(.+)<br>decks<\/a>$/
    );

    this.archetype = {
      id: parseInt(archetypeId, 10),
      name: archetypeName.trim(),
    };

    this.player = $('table .chosen_tr [align=right] .topic').text().trim();

    this.result = $('table .chosen_tr [align=center]').text().trim();

    let currentGroup: CardGroup = 'mainboard';
    $('table table table td td').each((_, el) => {
      const isCard = $('.hover_tr', el).length > 0;
      if (isCard) {
        this[currentGroup].push(parseCard($(el)));
      } else {
        currentGroup = parseGroup($(el));
      }
    });
  }
}

const parseCard = ($el: Cheerio): WithCount<Card> => {
  const text = $el.text();
  const match = text.match(/^(\d+) (.+)$/);
  if (!match) throw new Error(`Cannot parse card: ${text}`);

  const count = parseInt(match[1], 10);
  const name = match[2].trim();
  const isCompanion = $el.find('[id="companion_%%%"]').length > 0;

  const card: WithCount<Card> = { count, name };
  if (isCompanion) card.isCompanion = true;

  return card;
};

const parseGroup = ($el: Cheerio): CardGroup => {
  switch ($el.text()) {
    case 'COMMANDER':
      return 'commander';
    case 'SIDEBOARD':
      return 'sideboard';
    default:
      return 'mainboard';
  }
};
