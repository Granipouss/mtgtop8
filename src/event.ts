import { getPage, url, match, parseDeckLink, isDefined } from './utils';
import { ShortFormat } from './format';
import { ShortDeck } from './deck';

export type ShortEvent = { id: number; name: string };

type WithRanks<T> = T & { ranks: number[] };

export class Event implements ShortEvent {
  public static async get(id: number): Promise<Event> {
    const $ = await getPage(url.event(id));

    if ($.html().includes('No event could be found.')) {
      throw new Error(`No event ${id}`);
    }

    return new Event($);
  }

  id: number;
  name: string;

  format: ShortFormat;
  rank: number;
  playerCount: number;
  date: Date;
  decks: WithRanks<ShortDeck>[] = [];

  private constructor($: CheerioStatic) {
    const chosenLinkData = parseDeckLink($('.chosen_tr a').attr('href'));

    this.id = chosenLinkData.event;

    this.name = $('.w_title td:first-child').text().trim();

    this.format = {
      name: match($('table table td[align="center"]').html() || '', /(.+) <div/)[0].trim(),
      code: chosenLinkData.format,
    };

    const stars = $('img[src="/graph/star.png"]').length;
    const bigstars = $('img[src="/graph/bigstar.png"]').length;
    this.rank = 10 * bigstars + stars;

    const [playerCount, date] = match($('table table td[align="center"]').html() || '', /(\d+) players - (.+)/);

    this.playerCount = parseInt(playerCount, 10);

    this.date = new Date(`${date}Z`);

    this.decks = $('.hover_tr, .chosen_tr')
      .get()
      .map((el) => {
        try {
          const linkData = parseDeckLink($(el).find('a').first().attr('href'));
          return {
            ranks: $(el).find('div:first-child').text().split('-').map(Number),
            eventId: linkData.event,
            id: linkData.deck,
            name: $(el).find('a').first().text(),
          };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }
}
