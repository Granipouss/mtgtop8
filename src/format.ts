import { getPage, match, ListOf, isDefined, url } from './utils';

import { ShortArchetype } from './archetype';
import { ShortEvent } from './event';

export type ShortFormat = { name: string; code: string };

export class Format implements ShortFormat {
  public static async get(code: string): Promise<Format> {
    const $ = await getPage(url.format(code));

    if (getTableRows($, /Metagame Breakdown/).length === 1) {
      throw new Error(`No format for code ${code}`);
    }

    return new Format($);
  }

  public static async getAll(): Promise<ShortFormat[]> {
    const $ = await getPage(url.base());
    return [...$('table').eq(1).find('td').get(), ...$('#other_tooltip_content div').get()]
      .map((el) => {
        try {
          const [code] = match($(el).find('a').attr('href'), /https:\/\/www\.mtgtop8\.com\/format\?f=(.+)/);

          const alt = $(el).find('img').attr('id');
          if (!alt) return;
          const name = alt
            .replace(/^menu_/, '')
            .replace(/_/g, ' ')
            .toLowerCase();

          return { name, code };
        } catch (error) {
          return;
        }
      })
      .filter(isDefined);
  }

  public name: string;
  public code: string;

  public archetypes: ListOf<ShortArchetype> = [];

  private constructor($: CheerioStatic) {
    this.code = match($('table table .hover_tr a').attr('href'), /archetype\?a=.+&meta=.+&f=(.+)/)[0];

    this.name = $('table table a[target=_blank]')
      .text()
      .replace(/ Construction Rules$/, '')
      .toLowerCase()
      .trim();

    this.archetypes = getTableRows($, /Metagame Breakdown/)
      .filter((el) => $(el).is('.hover_tr'))
      .map((el) => {
        const [idString] = match($(el).find('a').attr('href'), /archetype\?a=(\d+)&meta=.+&f=.+/);
        return {
          id: parseInt(idString, 10),
          name: $(el).find('a').text().trim(),
          count: parseInt($(el).find('[align=right]').first().text(), 10),
        };
      });
  }

  async getEvents(page = 1): Promise<ShortEvent[]> {
    const $ = await getPage(url.format(this.code, page));
    return getTableRows($, /^(Events \d+ to \d+|Last \d+ events)$/).map((el) => {
      const [idString] = match($(el).find('a').attr('href'), /event\?e=(\d+)&f=.+/);
      return {
        id: parseInt(idString, 10),
        name: $(el).find('a').text().trim(),
      };
    });
  }
}

const getTableRows = ($: CheerioStatic, title: RegExp): CheerioElement[] => {
  try {
    const table: CheerioElement | undefined = $('table table')
      .get()
      .find((el) => $('tr:first-child', el).text().match(title));
    return table ? $(table).find('tr').get().splice(1) : [];
  } catch (error) {
    return [];
  }
};
