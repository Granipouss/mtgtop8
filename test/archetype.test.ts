import { Archetype } from '../src/archetype';

describe('Archetypes', () => {
  it('should match snapshot for single', async () => {
    const data = await Archetype.get(214);
    expect(data).toMatchSnapshot();
  });

  it('should match snapshot with page', async () => {
    const data = await Archetype.get(226, 2);
    expect(data).toMatchSnapshot();
  });

  // it('should throw on not found', async () => {
  //   await expect(getDeck(0, 1)).rejects.toThrow('No deck 1 for event 0');
  // });
});
