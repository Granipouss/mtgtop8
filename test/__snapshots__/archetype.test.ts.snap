// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Archetypes should match snapshot for single 1`] = `
Archetype {
  "decks": Array [
    Object {
      "eventId": 26573,
      "id": 405134,
      "name": "Weenie White",
    },
    Object {
      "eventId": 26360,
      "id": 401188,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 24053,
      "id": 367839,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 23641,
      "id": 363959,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 23267,
      "id": 360094,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 23132,
      "id": 358956,
      "name": "Soul Sisters",
    },
    Object {
      "eventId": 22987,
      "id": 357571,
      "name": "Soul Sisters",
    },
    Object {
      "eventId": 22974,
      "id": 357450,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 22912,
      "id": 356769,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 22917,
      "id": 356828,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 22810,
      "id": 355898,
      "name": "Martyr Proc",
    },
    Object {
      "eventId": 22727,
      "id": 355188,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 22637,
      "id": 354458,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 22657,
      "id": 354615,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 22602,
      "id": 354082,
      "name": "Soul Sisters",
    },
    Object {
      "eventId": 22666,
      "id": 354687,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 22327,
      "id": 351489,
      "name": "Martyr Proc",
    },
    Object {
      "eventId": 22231,
      "id": 350560,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 21645,
      "id": 345345,
      "name": "Martyr Life",
    },
    Object {
      "eventId": 21582,
      "id": 344907,
      "name": "Martyr Life",
    },
  ],
  "id": 214,
  "name": "Martyr Life",
}
`;

exports[`Archetypes should match snapshot with page 1`] = `
Archetype {
  "decks": Array [
    Object {
      "eventId": 26475,
      "id": 403261,
      "name": "Burn",
    },
    Object {
      "eventId": 26492,
      "id": 403428,
      "name": "Red Deck Wins",
    },
    Object {
      "eventId": 26492,
      "id": 403425,
      "name": "Prowess Deck Wins",
    },
    Object {
      "eventId": 26499,
      "id": 403478,
      "name": "Prowess Deck Wins",
    },
    Object {
      "eventId": 26379,
      "id": 401504,
      "name": "Prowess Deck Wins",
    },
    Object {
      "eventId": 26379,
      "id": 401505,
      "name": "Burn",
    },
    Object {
      "eventId": 26360,
      "id": 401192,
      "name": "Burn",
    },
    Object {
      "eventId": 26276,
      "id": 399971,
      "name": "Red Deck Wins",
    },
    Object {
      "eventId": 26274,
      "id": 399921,
      "name": "Prowess Deck Wins",
    },
    Object {
      "eventId": 26221,
      "id": 399251,
      "name": "Burn",
    },
    Object {
      "eventId": 26314,
      "id": 400559,
      "name": "Burn",
    },
    Object {
      "eventId": 26121,
      "id": 397393,
      "name": "Red Deck Wins",
    },
    Object {
      "eventId": 26143,
      "id": 397745,
      "name": "Burn",
    },
    Object {
      "eventId": 26103,
      "id": 397109,
      "name": "Burn",
    },
    Object {
      "eventId": 26057,
      "id": 396169,
      "name": "Prowess Deck Wins",
    },
    Object {
      "eventId": 26056,
      "id": 396167,
      "name": "Mono Red Prowess",
    },
    Object {
      "eventId": 26035,
      "id": 395892,
      "name": "Burn",
    },
    Object {
      "eventId": 25952,
      "id": 394650,
      "name": "Burn",
    },
    Object {
      "eventId": 26107,
      "id": 397163,
      "name": "Naya Burn",
    },
    Object {
      "eventId": 25904,
      "id": 393844,
      "name": "RDW",
    },
  ],
  "id": 226,
  "name": "Red Deck Wins",
}
`;
